### PR TITLE
Update fasta code

### PR DESCRIPTION
**Pull request by jeromekelleher**
Update the FASTA output to use the new alignments method.

Stacked on #1825

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Ready for review - can you have a look please @hyanwong, most of this is your code originally I think.

**Comment by jeromekelleher** [comment]
Hmmm, this is complicated by #1893.

File: python/tests/test_fasta.py [hunk 1]
```diff
@@ -23,28 +23,91 @@
...
-    ts = msprime.simulate(
-        sample_size=10, length=length, mutation_rate=1e-2, random_seed=123
+@functools.lru_cache(maxsize=100)
+def create_data(sequence_length):
+    ts = msprime.sim_ancestry(
+        samples=5, sequence_length=sequence_length, random_seed=123
     )
-    ts = tsutil.jukes_cantor(ts, length, 1.0, seed=123)
-    assert ts.num_sites == length
+    ts = msprime.sim_mutations(ts, rate=0.1, random_seed=1234)
```
Comment (hyanwong) [inline python/tests/test_fasta.py:45]: Nice to switch to sim_mutations now: 👍

**Comment by hyanwong** [comment]
> Ready for review - can you have a look please @hyanwong, most of this is your code originally I think.

Pending #1893 this LGTM. Two things that I think people will want to be able to do are:

1. Only output a few of the haplotypes, or a short section of the sequence. This can be done, of course, using `.simplify()` in the first instance and `.keep_intervals` in the second, but it would be good to show people how to do this, either in a tutorial and/or mentioning it in the docstring. There is a (weak) argument for a parameter that says which samples to output, as `simplify()` will change the node numbers, meaning the labels might not be what is expected. I imagine people might do e.g. `my_ts.simplify([4, 5, 6]).as_fasta()` and expect the labels to be [`n4`, `n5`, `n6`] rather than [`n0`, `n1`, `n2`]. Something like `my_ts.simplify().as_fasta(samples=[4, 5, 6]) would allow us to output a string with [`n4`, `n5`, `n6`] as labels. But it may not be worth adding to the complexity of the  as_fasta/write_fasta functions, if we can just warn people to be careful
2. Have a human-readable label for the FASTA strings (was previously provided by the `sequence_ids` parameter). This would reduce the problems in 1., as I assume most people would label using metadata.

One solution to 2., which I don't think this should be part of this PR, but may be worth considering, is to have a general labelling class that could also be used in other output routines, such as `draw_svg`. E.g. one or more of the following:

```
# as a function
ts.write_fasta(node_labels=NodeLabels(lambda n: n.metadata['name'])
# or as a mapping
ts.write_fasta(node_labels=NodeLabels({n.id: n.metadata['name'] for n in ts.nodes()})
# or as a simple array
ts.write_fasta(node_labels=NodeLabels([n.metadata['name'] for n in ts.nodes()])
```

**Comment by jeromekelleher** [comment]
Thanks @hyanwong.

1. We should definitely do this at some point, but the functionality needs to be propagated back through ``.alignments`` etc first, and that's some work.
2. Adding custom labels like we have for ``as_newick`` would definitely be useful, but again the ability to pick specific samples for alignments would be necessary to make this work in general. For now, it's more important that the default node labelling is simple and consistent across various export functions (#1817).

**Comment by jeromekelleher** [comment]
I'm going to close this in favour of #1894, as it'll be easier to review the fasta code as "new" when fixing #1893

**Pull request closed without merging**